src/libs/maintaining-std.md

```diff
@@ -0,0 +1,132 @@
+
+The rules around what's sound and what's not can be subtle. See the [Unsafe Code Guidelines WG] for current thinking, and consider pinging `@rust-lang/libs` and/or somebody from the WG if you're in any doubt.
+
+### Is that `#[inline]` right?
+
+Inlining is a trade-off between potential execution speed, compile time and code size.
+
+You should add `#[inline]`:
+
+- To public, small, non-generic functions.
+
+You shouldn’t need `#[inline]`:
+
+- On methods that have any generics in scope.
+- On methods on traits that don’t have a default implementation.
+- On `const` items.
+
+#### What about `#[inline(always)]`?
+
+You should just about never need `#[inline(always)]`. It may be beneficial for private helper methods that are used in a limited number of places or for trivial operators. A micro benchmark should justify the attribute.
+
+### Is there any potential breakage?
+
+Breaking changes should be avoided when possible. [RFC 1105] lays the foundations for what constitutes a breaking change. Breakage may be deemed acceptable or not based on its actual impact, which can be approximated with a `crater` run.
+
+For changes where the value is high and the impact is high too, there are strategies for minimizing the impact:
+
+- Using compiler lints to try phase out broken behavior.
+
+### Could this affect inference?
+
+New impls on public traits for public items may cause inference to break when there are generics involved.
+
+#### Are there `#[fundamental]` items involved?
+
+Blanket trait impls can't be added to `#[fundamental]` types because they have different coherence rules. That includes:
+
+- `Box<T>`
+- `Pin<T>`
+
+Also see [RFC 1023] for details.
+
+### Is there specialization involved?
+
+We try to avoid leaning on specialization too heavily, limiting its use to optimizing specific implementations. Any use of specialization that changes how methods are dispatched for external callers should be carefully considered.
+
+### Does this change drop order?
+
+Changes to collection internals may affect the order their items are dropped in. This has been accepted in the past, but should be noted.
+
+### Could `mem::replace` break assumptions?
+
+Any value behind a `&mut` reference can be replaced with a new one.
+
+### Could `mem::forget` break assumptions?
+
+Rust doesn't guarantee destructors will run, so code should avoid relying on them for safety. Remember, [everybody poops][Everybody Poops].
+
+### Is the commit log tidy?
+
+PRs shouldn’t have merge commits in them. If they become out of date with `master` then they need to be rebased.
+
+## Merging PRs
+
+PRs to `rust-lang/rust` aren’t merged manually using GitHub’s UI or by pushing remote branches. Everything goes through [`bors`].
+
+### When you’re confident it’ll build
+
+Consider explicitly specifying `rollup`.
```

Comment:
Conversely, if it might regress perf, consider `rollup=never`

Reply:
@Centril As somebody who does a lot of the rollup work how would you determine whether to `rollup` or not? Is there anything you wish reviewers did consistently that would make that work smoother?

Reply:
We recently added a guide on rollups at https://forge.rust-lang.org/release/rollups.html.

For libs-related PRs, rolling up is usually fine, in particular if it's only a new unstable addition or if it only touches docs (with the exception of intra doc links which complicates things while the feature has bugs...).

Sometimes a libs PR might update some submodule and such and then I'd use more caution and probably not use `@bors rollup`.

For libs stuff, it's probably only for things which supposedly may improve compile-time perf which I'd use `rollup=never` for. https://github.com/rust-lang/rust/pull/63061 would be an example of this.

Reply:
Do you mind if I pretty much just drop that comment in there?

Reply:
Not at all, but perhaps replace "I'd" with "we'd" or something less confusing. ;)